srml/balances/src/lib.rs

```diff
@@ -1040,12 +1042,18 @@ where
 }
 
 impl<T: Trait<I>, I: Instance> MakePayment<T::AccountId> for Module<T, I> {
-	fn make_payment(transactor: &T::AccountId, encoded_len: usize) -> Result {
-		let encoded_len = T::Balance::from(encoded_len as u32);
-		let transaction_fee = Self::transaction_base_fee() + Self::transaction_byte_fee() * encoded_len;
+	fn make_payment(transactor: &T::AccountId, weight: Weight) -> Result {
+		let VARIABILITY_FEE = Permill::from_parts(40);
+		let VARIABILITY_FEE_SQUARED = Permill::from_parts(16);
+		let potential_weight = <system::Module<T>>::all_extrinsics_weight() + weight;
+		let saturation_diff = potential_weight - IDEAL_TRANSACTIONS_WEIGHT;
+		let first_term = VARIABILITY_FEE * saturation_diff; // TODO: divide by 1000000 * 10
+		let second_term = VARIABILITY_FEE_SQUARED * saturation_diff * saturation_diff / 2; // TODO: ^ditto
+		let fee_multiplier = 1 + first_term + second_term;
+		let transaction_fee = weight * fee_multiplier;
 		let imbalance = Self::withdraw(
 			transactor,
-			transaction_fee,
+			T::Balance::from(transaction_fee),
```

Comment:
probably works also like
```suggestion
			transaction_fee.into(),
```